miniconf_derive/src/lib.rs

```diff
@@ -222,6 +230,27 @@ fn derive_struct(
 
                 meta
             }
+
+            fn traverse_by_key<P, F, E>(
+                keys: &mut P,
+                mut func: F,
+            ) -> miniconf::Result<E>
+            where
+                P: Iterator,
+                P::Item: miniconf::Key,
+                F: FnMut(miniconf::Ok, usize, &str) -> Result<(), E>,
+            {
+                match keys.next() {
+                    None => Ok(miniconf::Ok::Internal(0)),
+                    Some(key) => {
+                        let index = miniconf::Key::find::<Self>(key).ok_or(miniconf::Error::NotFound(1))?;
+                        match index {
+                            #(#traverse_by_key_arms ,)*
+                            _ => Err(miniconf::Error::NotFound(1)),
```

Comment:
I'm not really following why the indices are being used. I also am concerned about consistency. Can we maybe add an impl to `Error` and make these use a default-style constructor, since the various types are always newly constructed with the same initial value? Something like:

```rs
miniconf::Error::not_found()
```

The idea being that the index provided is constant. I'm not sure if this is necessarily true, but in general it's really hard for me to understand what these indices are supposed to be initialized at, and would help if we didn't need to think about the complexity everywhere.

Reply:
I'd like to be explicit here. We are at depth = 1. This needs to be there because we got `Some()` from `next()`. It's distinct from the `Internal(0)` above. If we had `not_found()` and `internal()` that would obscure it for me.